File: 206-reverse-linked-list/206-reverse-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* reverseList(struct ListNode* head) {
    struct ListNode *cur;
    struct ListNode *prev;
    struct ListNode *next;
    struct ListNode prehead;

    if (!head) {
        return NULL;
    }
    
    prehead.next = head;
    cur = &prehead;
    prev = cur;
    next = cur->next;

    while (next) {
        cur = next;
        next = cur->next;
        cur->next = prev;
        prev = cur;
    }
    
    head->next = NULL;
    
    return cur;
}

File: 206-reverse-linked-list/206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return None
        
        prev = head
        cur = head.next
        
        while cur:
            next_node = cur.next
            cur.next = prev
            prev = cur
            cur = next_node
        
        head.next = None
        
        return prev